test/test_optim.py

```diff
@@ -29,6 +31,78 @@ def test_errors(self, device, dtype, optim_info):
                 raise NotImplementedError(f"Unknown error type {error_input.error_on}")
 
 
+    def _test_derived_optimizers(self, device, dtype, optim_info, flag):
+        assert flag in ("foreach", "fused")
+
+        # why 7? iteration 7 is where we start to see differences for RAdam
+        # params interacting with the small eps value, because that's right
+        # after rho_t becomes greater than 5 in step 6.
+        kIterations = 7
+
+        optim_inputs = optim_info.optim_inputs_func()
+        optim_cls = optim_info.optim_cls
+        for optim_input in optim_inputs:
+            updated_params, state = [], []
+            kwargs = deepcopy(optim_input.kwargs)
+            if (kwargs.get("capturable", False) and
+               (str(device) == "cpu" or optim_cls.__name__ == "ASGD")):
+                # capturable is not supported on CPU nor in single tensor ASGD
+                continue
+            for flag_value in (False, True):
+                kwargs[flag] = flag_value
+                input = torch.tensor(
+                    [0.1, 0.2, 0.3, 0.4, 0.5, 0.6], dtype=dtype, device=device
+                ).reshape(3, 2)
+
+                torch.manual_seed(1)
+                model = torch.nn.Sequential(
+                    torch.nn.Linear(2, 3),
+                    torch.nn.Sigmoid(),
+                    torch.nn.Linear(3, 1),
+                    torch.nn.Sigmoid(),
+                )
+                model.to(dtype=dtype, device=device)
+
+                # foreach/fused optimizers should be tested with a
+                # zero_size tensor as its last param.
+                # ref: https://github.com/pytorch/pytorch/issues/100701
+                empty_params = [torch.empty((), device=device, dtype=dtype)]
+                params = list(model.parameters()) + empty_params
+
+                optimizer = optim_cls(params, **kwargs)
+
+                for i in range(kIterations):
+                    optimizer.zero_grad()
+                    output = model(input)
+                    loss = output.sum()
+                    loss.backward()
+
+                    # Test that step behaves as expected (a no-op) when grads are set to None
+                    if i == 0:
```

Comment:
nit: maybe i == 3 would be better? If this is the very first iteration, there are no momentum buffers anyways. So we wouldn't see the difference?